library/core/src/main/java/com/google/android/exoplayer2/extractor/ts/Ac3Reader.java

```diff
@@ -188,18 +192,36 @@ private boolean skipToNextSync(ParsableByteArray pesBuffer) {
   private void parseHeader() {
     headerScratchBits.setPosition(0);
     Ac3Util.Ac3SyncFrameInfo frameInfo = Ac3Util.parseAc3SyncframeInfo(headerScratchBits);
-    if (format == null || frameInfo.channelCount != format.channelCount
-        || frameInfo.sampleRate != format.sampleRate
-        || frameInfo.mimeType != format.sampleMimeType) {
+    if ((format == null || frameInfo.sampleRate != format.sampleRate
+        || frameInfo.mimeType != format.sampleMimeType)
+        && (frameInfo.streamType == Ac3Util.STREAM_TYPE_TYPE0
+        || frameInfo.streamType == Ac3Util.STREAM_TYPE_UNDEFINED)) {
       format = Format.createAudioSampleFormat(trackFormatId, frameInfo.mimeType, null,
           Format.NO_VALUE, Format.NO_VALUE, frameInfo.channelCount, frameInfo.sampleRate, null,
           null, 0, language);
       output.format(format);
     }
+
+    // If the (E-)AC-3 stream contains I0 + D0, and I0.channels = 6, D0.channels = 4,
+    // then it's 7.1-channel (E-)AC-3 stream.
+    // TODO: handle other (E-)AC-3 streams?
+    if (maybeReconfigure == true && frameInfo.streamType == Ac3Util.STREAM_TYPE_TYPE1
+        && format.channelCount == 6 && frameInfo.channelCount == 4) {
+      format = Format.createAudioSampleFormat(trackFormatId, frameInfo.mimeType, null,
+          Format.NO_VALUE, Format.NO_VALUE, 8, frameInfo.sampleRate, null,
+          null, 0, language);
+      output.format(format);
+      maybeReconfigure = false;
+    }
     sampleSize = frameInfo.frameSize;
     // In this class a sample is an access unit (syncframe in AC-3), but the MediaFormat sample rate
     // specifies the number of PCM audio samples per second.
     sampleDurationUs = C.MICROS_PER_SECOND * frameInfo.sampleCount / format.sampleRate;
+    // If the (E-)AC-3 stream is 7.1-channel, the dependent sub-stream shares the sampleDuration
+    // with independent sub-stream.
+    if (frameInfo.streamType == Ac3Util.STREAM_TYPE_TYPE1) {
+      sampleDurationUs = 0;
```

Comment:
Dolby's 7.1 EAC3 stream consists of I(Independent frame) and D(dependent frame). These two frames cross appear in the stream, i.e. I,D,I,D,......
One I frame and D frame is an access unit, which is a basic unit for decoder. ( For example, for 7.1ch EAC3 stream, I frame includes L,R,C,LFE,Ls',Rs' channels. Decoder drops Ls' and Rs' channel data since they are used for 5.1 downmix output. D frame includes Ls,Rs,Lb,Rb. The final output is L,R,C,LFE,Ls,Rs,Lb,Rb (total 8 channels) ) 
When first I frame occurs, sampleDurationUs is calculated. 
Then first D frame occurs, sampleDurationUs is 0 since D frame doesn't increase more samples. It just add more channels for previous I frame. I+D is a whole access unit from decoder perspective.

I hope it explains the code above. :)